pkg/cache/bigquerycache/bigquery.go

```diff
@@ -0,0 +1,302 @@
+)
+
+var (
+	persistentCacheWarmMissMetric = promauto.NewCounterVec(prometheus.CounterOpts{
+		Name: persistentCacheWarmMiss,
+		Help: "Number of persistent cache gets that were not in the warm cache.",
+	}, []string{})
+	persistentCacheBackendMissMetric = promauto.NewCounterVec(prometheus.CounterOpts{
+		Name: persistentCacheBackendMiss,
+		Help: "Number of persistent cache gets that were not in the backend cache.",
+	}, []string{})
+	persistentCacheGetMetric = promauto.NewCounterVec(prometheus.CounterOpts{
+		Name: persistentCacheGet,
+		Help: "Number of persistent cache get requests.",
+	}, []string{})
+	persistentCacheBackendGetMetric = promauto.NewCounterVec(prometheus.CounterOpts{
+		Name: persistentCacheBackendGet,
+		Help: "Number of persistent cache get that query the backend cache.",
+	}, []string{})
+	persistentCacheReadOnlySetMetric = promauto.NewCounterVec(prometheus.CounterOpts{
+		Name: persistentCacheReadOnlySet,
+		Help: "Number of persistent cache set calls for a read only client.",
+	}, []string{})
+	persistentCacheBackendSetMetric = promauto.NewCounterVec(prometheus.CounterOpts{
+		Name: persistentCacheBackendSet,
+		Help: "Number of persistent cache backend set calls.",
+	}, []string{})
+)
+
+// Cache implementation that supports storing data to a bigquery table
+// Introduced to handle large data structures that are costly in time and $$
+// as well as considered largely static, allowing for longer storage between queries
+//
+// Additionally the concept of the cache being read only supports separate processes
+// writing cached data and reading.  In a case where one process will write the cache data
+// it should have a shorter maxExpiration duration in the Cache initialization via input parameter.
+// This duration is separate from the Duration passed in the Set call
+// and can be used to cause new entries to be written to the backend table before they are considered expired.
+// If a cache item has a 7 day duration passed in on the Set call
+// and the read only cache process is configured with a maximum 10 day duration for all cache items
+// the process that writes to the cache could be configured with a 1 (or more) day maximum duration that would cause
+// it to ignore any entries older than 24 hours and perform a new query and update the cache once a day.
+// Providing regular cache updates to the read only process and allowing for outages as well.
+//
+// minExpiration is the minimum duration that will be written to the bigquery cache table
+// items that are cached with a Set duration shorter than minExpiration will skip the biqquery
+// caching and rely on the main caching layer.  The minExpiration is also used to determine if a Get
+// that misses the main caching layer should check the bigquery cache table.  If the duration is below the min
+// it will not check for an entry but instead return a miss.
+type Cache struct {
+	client        *sippybq.Client
+	readOnly      bool
```

Comment:
Wondering if we need an inverse of this that is `primaryWrite` indicating skip the warm cache lookup and go to the backend cache.  In the case above where maxExpiration is shortened for the cache write process that won't impact the read from the warm cache which would lead to it not considering the data expired and therefore not refreshing the cache as expected.